### PR TITLE
CLI: Improve `add` command & add tests

### LINKED ISSUE #25871: [Bug]: Storybook add with version selectors
### Describe the bug

`storybook add` installs addons in your `package.json`, `node_modules`, and `main.js` config.

It currently fails with version specifiers:

```
shilman@MBP14 chromatic-test-1 % npx storybook add @chromatic-com/storybook@1.0.3--canary.171.b614185.0
Verifying @chromatic-com/storybook@1.0.3--canary.171.b614185.0
Installing @chromatic-com/storybook@1.0.3--canary.171.b614185.0@^1.0.3--canary.171.b614185.0

An error occurred while installing dependencies:
NPM error EINVALIDTAGNAME
```

### To Reproduce

_No response_

### System

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ndelangen**
Closes https://github.com/storybookjs/storybook/issues/25871

## What I did

I found that the add command contained a few quicks and bugs that I fixed.

- when adding a single digit version such as this: `sb add package@1` it would install the latest version instead
- when adding a storybook package it wouldn't warn you when installing a specific version that isn't the current
- the logic to split a package+version specifier into a packageName & versionSpecifier wasn't setup to deal with scope packages

I added tests for the functionality of the `add` command.
The `getVersionSpecifier` function is tested separately, because regexes can be tricky, and it's faster to iterate on it this way if there is ever another bug with in.

I can remove the warning functionality if it isn't desired.
I figured i'd be nice to have because it would prevent it from later becoming an issue the user would have to fix with the doctor command.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

I've tested this manually in a sandbox via running the CLI directly from dist:

```
(next)⚡ [7] % node ../../code/lib/cli/bin/index.js add @chromatic-com/storybook@1.0.0
/Users/me/Projects/Storybook/mono/storybook/code/lib/cli/bin/index.js:23
  throw error;
  ^

Error: Addon @chromatic-com/storybook is already installed; we skipped adding it to your .storybook/main.ts.
    at add (/Users/me/Projects/Storybook/mono/storybook/code/lib/cli/dist/generate.js:3699:11)

Node.js v18.19.1
(next)⚡ [7] % node ../../code/lib/cli/bin/index.js add @storybook/addon-onboarding@1.0.0
/Users/me/Projects/Storybook/mono/storybook/code/lib/cli/bin/index.js:23
  throw error;
  ^

Error: Addon @storybook/addon-onboarding is already installed; we skipped adding it to your .storybook/main.ts.
    at add (/Users/me/Projects/Storybook/mono/storybook/code/lib/cli/dist/generate.js:3699:11)

Node.js v18.19.1
(next)⚡ [7] % node ../../code/lib/cli/bin/index.js add @storybook/addon-measure@1.0.0   
Verifying @storybook/addon-measure
Installing @storybook/addon-measure@^1.0.0
Adding '@storybook/addon-measure@1.0.0' to main.js addons field.
(next)⚡ % node ../../code/lib/cli/bin/index.js add @storybook/addon-measure      
Verifying @storybook/addon-measure
Installing @storybook/addon-measure@^8.0.0-rc.1
Adding '@storybook/addon-measure' to main.js addons field.
(next)⚡ %        
```

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: code/lib/cli/src/postinstallAddon.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+      console.log(`Running postinstall script for ${addonName}`);
+      await postinstall(options);
+    } catch (e) {
+      console.error(`Error running postinstall script for ${addonName}`);
+      console.error(e);
+    }
+  } catch (e) {
+    // no postinstall script
+  }
+};
```
Comment (ndelangen) [inline code/lib/cli/src/postinstallAddon.ts:19]: I extracted this, to be able to easily mock it in my test.

File: code/lib/cli/src/add.ts [hunk 2]
```diff
@@ -71,9 +61,11 @@ const checkInstalled = (addonName: string, main: any) => {
  */
 export async function add(
   addon: string,
-  options: { packageManager: PackageManagerName; skipPostinstall: boolean }
+  options: { packageManager: PackageManagerName; skipPostinstall: boolean },
+  logger = console
```
Comment (ndelangen) [inline code/lib/cli/src/add.ts:65]: Added this extra optional parameter to get less noisy tests

File: code/lib/cli/src/add.ts [hunk 3]
```diff
@@ -85,43 +77,52 @@ export async function add(
...
+  let version = inputVersion;
+
+  if (!version && isCoreAddon(addonName) && storybookVersion) {
+    version = storybookVersion;
+  }
+  if (!version) {
+    version = await packageManager.latestVersion(addonName);
+  }
+
+  if (storybookVersion && version !== storybookVersion) {
```
Comment (valentinpalkovic) [inline code/lib/cli/src/add.ts:105]: Should we show the warning message only for Storybook Core Addons, which should have the same version installed as the other Storybook monorepo packages? This check shouldn't be applied for e.g. `npx storybook@next add @storybook/addon-styling-webpack`, where the version of `@storybook/addon-styling-webpack` will not match with the installed Storybook version.

File: code/lib/cli/src/add.test.ts [hunk 4]
```diff
@@ -104,7 +104,18 @@ describe('add', () => {
...
+      MockedConsole
+    );
+
+    expect(MockedConsole.warn).not.toHaveBeenCalledWith(
+      expect.stringContaining(`is not the same as the version of Storybook you are using.`)
+    );
+  });
+  test('not warning when installing unrelated package', async () => {
+    await add(
+      '@storybook/addon-docs',
```
Comment (valentinpalkovic) [inline code/lib/cli/src/add.test.ts]: This test does the same as the one in line 105. I guess you want to use a non-core addon in line 118.

**Review by valentinpalkovic — APPROVED** [review]
Code-wise LGTM!

**Pull request merged**